Migrated HLSL 2016 with SM 6_0

toddmacedo opened the pull request.

toddmacedo committed:
Migrated HLSL 2016 with SM 6_0

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-landing-page.md:
```diff
@@ -0,0 +1,9 @@
+# High-level shader language (HLSL)
```
Let's describe and link the sub-headings under this landing page in the TOC.  Let's do the same for the section pages (adding pages if necessary).  We can follow the example of the legacy HLSL landing/section pages.

The goal is to help people find the documentation they are looking for by summarizing the sections and the types of content found within them.  Point out when certain types of content divisions might not be immediately obvious.  Such as how we place certain things under the built-in section rather than the language section, when it's not immediately obvious where they might be found.

toddmacedo replied:
Done.

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-language-access_qualifiers.md:
```diff
@@ -0,0 +1,3 @@
+# Access Qualifiers
+
+See [`DirectX Spec`](https://microsoft.github.io/DirectX-Specs/) for more details.
```
What's this page for?  We already have: `hlsl-built-ins-raytracing-payload-access-qualifiers.md`.  This just links to the top-level spec repo, so it's very unhelpful.  This page is probably lefover and should be deleted.

toddmacedo replied:
For the time being, there are only raytracing payload access qualifiers, so I removed this section from the repo.

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-language-attributes.md:
```diff
@@ -0,0 +1,3 @@
+# Attributes
+
+Attributes describe additional behavior for language features.  These attributes precede language features and are enclosed in square brackets.
```
I don't think this is nearly sufficient for attributes.  Ideally, we should break down the different attribute types that HLSL supports, show the syntax (that's the language part), and link to the built-in section for the set of attributes recognized in each location.

For instance:
- Function Attributes: before a function declaration/definition
- Loop Attributes: before a for/do/while statement
- Branch Attributes: before an if statement
- Switch Attributes: before a switch statement
- etc...

toddmacedo replied:
Alright, wrote original copy and headers for these sections.  I expanded flow control into loop attributes and branch attributes.

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-built-ins-functions_math_logic.md:
```diff
@@ -0,0 +1,3332 @@
+# Math and Logic Operations
+
+### abs
```
If all these are going to be described on the same page, we at least need TOC nodes for navigation.  It's also visually hard to parse and see where one operation ends and another begins.  I wish the content under an operation was all indented or something, so it would be visually obvious.

tex3d replied:
Could parameter, return value and type tables be merged into one table by simply adding the description to the type table?

I imagine this would need to be done with a script, otherwise it's a lot of mechanical labor.

toddmacedo replied:
I could write a simple C# program with regular expressions.  It won't be too complicated.

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-language-functions.md:
````diff
@@ -0,0 +1,350 @@
+## Function Declaration Syntax
+
+```syntax
+[StorageClass] [clipplanes()] [precise] Return_Value Name ( [ArgumentList] ) [: Semantic] {   [StatementBlock] };
````
`[clipplanes()]` should be replaced with `[FunctionAttributes]`.

toddmacedo replied:
also added SV_ClipDistance to System Value semantics.

toddmacedo replied:
Done.  Put clipplanes in the built-in attributes and changed the clipplanes into attributes in the language section.

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-language-functions.md:
```diff
@@ -0,0 +1,350 @@
+
+Modifier that redefines a function declaration. `inline` is currently the only modifier value. The modifier value must be `inline` because it is also the default value. Therefore, a function is inline regardless of whether you specify `inline`, and all functions in HLSL are inline. An inline function generates a copy of the function body (when compiling) for each function call. This is done to de ...
+
+#### `Clipplanes`
```
This should be replaced with `FunctionAttributes`.  `[clipplanes()]` is just one function attribute valid on the vertex shader entry point specifically.  There are many more function attributes, which should be described under the built-ins section for the relevant contexts.

General attribute syntax should be shown here, since it's not clear from the above what these would look like.

We should link to a `FunctionAttributes` section on a general `HLSLAttributes` page describing the various kinds of attributes, linking to the built-in attributes for each kind.

The general syntax for HLSL attributes is something like:
```
HLSLAttributes : HLSLAttribute [ "," HLSLAttributes ]
HLSLAttribute : "[" Name [ "(" Arguments ")" ] "]"
Name : Identifier
Arguments : Argument [ "," Arguments ]
Argument : depends on the attribute
```
Though we don't use this sort of grammar convention for syntax in the docs - I'm just not sure how explicit `[` and `]` can be clearly shown with the syntax for function above.

toddmacedo replied:
There is some sort of error in the built-in attributes section.  I will put in clip planes and debug in the process.

toddmacedo replied:
Added attributes section linking to both the language definition as well as the built-in attributes section.

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-method-getDimensions.md:
````diff
@@ -0,0 +1,349 @@
+// None.
+```
+
+## Supported Shader Models
````
This table looks about right for Sample() and some related methods, but not for GetDimensions().  GetDimensions() should be supported at every stage on every shader model.

toddmacedo replied:
We should remove compatibility currently.

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-method-sample.md:
```diff
@@ -0,0 +1,226 @@
+| x | Supported |
+| 1* | Supported with optional feature: `SHADER_FEATURE_DERIVATIVES_IN_MESH_AND_AMPLIFICATION_SHADERS` |
+
+>TBD: Make optional feature into a link
```
This TBD should be resolved and removed from here.  It's replicated all over the place, along with the table that isn't correct for most other operations.

toddmacedo replied:
I think we should remove these compatibility tables until we have a good method of researching compatibility.

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-method-sample.md:
```diff
@@ -0,0 +1,226 @@
+      in  SamplerState  samp,
+      in  float3        coord
+   [, in  float         lodClamp
+   [, out uint          status ] ] ] );
```
typo: extra square close bracket `]`.  Same with next method signature.

toddmacedo replied:
Thank you.  Corrected.

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:
```diff
@@ -0,0 +1,283 @@
+      in  int2          offset,
+   [, in  int2          offset2]
+   [, in  int2          offset3]
+   [, in  int2          offset4]   
```
Offset2/3/4 aren't independently optional.  You can either use one offset, or 4 offsets.  So perhaps the syntax here should look more like this:
```
   [, in  int2          offset2
    , in  int2          offset3
    , in  int2          offset4]
```

tex3d replied:
Oh and same for all similar places here and for normal Gathers

toddmacedo replied:
I will also include in the introduction that the signatures are valid only with the first offset value or all 4.  There are no other signatures.

toddmacedo replied:
Added in your alternate text.

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:
```diff
@@ -0,0 +1,283 @@
+   [, in  int2          offset4]   
+   [, out uint          status ] );
+
+ElementType Texture2D<ElementType>::GatherCmpGreen(
```
Do we need to expand out each signature for Red, Green, Blue, and Alpha?  This is just component selection on identical method overloads.  Could the method name be written: `GatherCmp[Red|Green|Blue|Alpha]` with one signature instead?  (same for normal Gather)

toddmacedo replied:
I see what you are saying.  It seems like a good place to spell out everything is in the function signatures.  Of course, can easily correct if needed.

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:
```diff
@@ -0,0 +1,283 @@
+
+| Parameter | Description |
+| - | - |
+| in [`SamplerState samp`](#samplerstate-samp) | SamplerState object that determines how sampling will take place. |
```
This should be `SamplerComparisonState` for `GatherCmp*`.

toddmacedo replied:
Thank you. Done.

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:
```diff
@@ -0,0 +1,283 @@
+
+### `SamplerState samp`
+
+A Sampler state. This is an object declared in an effect file that contains state assignments.
```
SamplerComparisonState...

toddmacedo replied:
Done. Thanks.

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:
```diff
@@ -0,0 +1,283 @@
+A Sampler state. This is an object declared in an effect file that contains state assignments.
+
+> My alternate text:
+> SamplerState object that determines how sampling will take place.
```
Should have resolved these and "TBD" items instead of replicating them elsewhere.

toddmacedo replied:
Great alternate text here.

toddmacedo replied:
Replaced all instances with links added.

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:
```diff
@@ -0,0 +1,283 @@
+
+## Remarks
+
+None.
```
Nothing describes what Red/Green/Blue/Alpha means in the method name.  Wouldn't this be a suitable place for that?

toddmacedo replied:
I have added explanation for gather.

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-method-gatherCmp.md:
````diff
@@ -0,0 +1,283 @@
+## Example
+
+```HLSL
+// N/A
````
Should probably have an example if we have a section for it.

toddmacedo replied:
I agree.  If you can find an example for this method, please let me know.  For completeness, every section has an example even if it is blank.

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-built-ins-functions_math_logic.md:
```diff
@@ -0,0 +1,3332 @@
+
+
+
+| Name  | [**Template Type**](dx-graphics-hlsl-intrinsic-functions.md)                                                  | [**Component Type**](dx-graphics-hlsl-intrinsic-functions.md)                 | Size                           |
```
These links to `dx-graphics-hlsl-intrinsic-functions.md` aren't right.  Looks like this was replicated throughout this file.

toddmacedo replied:
Fixed.

tex3d commented on desktop-src/direct3dhlsl_current/hlsl-built-ins-functions_math_logic.md:
```diff
@@ -0,0 +1,3332 @@
+
+| Name  | [**Template Type**](dx-graphics-hlsl-intrinsic-functions.md)                                                  | [**Component Type**](dx-graphics-hlsl-intrinsic-functions.md)                 | Size                           |
+|-------|----------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------|--------------------------------|
+| *x*   | [**scalar**](dx-graphics-hlsl-intrinsic-functions.md), **vector**, or **matrix** | [**float**](/windows/desktop/WinProg/windows-data-types), [**int**](/windows/desktop/WinProg/windows-data-types) | any                            |
```
If types are linked, they should probably link to the appropriate HLSL type definitions, not windows-data-types, which doesn't even define these types, and instead defines a bunch of types from Windows headers that are not applicable to HLSL at all.  For now, I'd say just get rid of the links here.

toddmacedo replied:
Alright, fixed links to the proper locations in the current documentation.

toddmacedo committed:
Added landing pages.

toddmacedo commented:
Links Fixed:
•	HLSL Language
o	Flow Control
	Do statement
	For statement
	While statement
o	Functions
	Code sample link does not work in original as well.
	BasicHLSL10.fx
	AdvancedParticles.fx
	BasicHLSL10Sample
	ContentStreaming Sample
o	Statements
	Expressions
•	Variables goes to variable syntax
•	Added operator link
	Added additional
•	Variables
•	Operators
	Added math and logic links
•	All
•	Any
•	Asfloat
•	Asint
•	Asdouble
	Binary Casts
•	Asfloat
•	Asint
•	Asuint
	Bitwise operators
•	Fixed link to sm 4.0
	Comparison operators
•	Any
•	All
	Statement Blocks
o	User-defined types
	Enumerations
•	Code block at end
	Structs
•	HLSL packing rules redirect
•	Shader model 4 redirect
o	Attributes
	Entry function attributes link
	Branching attributes link
o	Packoffset
	Link to declaring a variable type
o	Register
	Link to shader profile
	Packoffset
o	Global Blocks
	Add link to register
	Add link to packoffset
	Added variable links
	Added vector links
	Added arrays
	Fixed DX11 links
o	Root signatures
	D3D_ROOT_SIGNATURE_FLAGS fixed
•	Built-ins
o	Preprocessor directives
	#define link
	Pragma directives
•	Missing 
o	Def
o	Message
o	Pack_matrix
o	Constants
	Added shader section.
o	Math and logic functions
	](intrinsic functions)
o	Raytracing functions
	Intersection Attributes structure
o	Attributes
	Link in intro
	SV_GSInstanceID
	ID3D11DeviceContext::Dispatch
o	Semantics
	Multiple links fixed.
•	Appendix
o	HLSL Language Versions
	Enum and enum class fixed.
o	Shader Model page
	Fixed links in high-level page
